In [1]:
import pandas as pd
import json
import plotly.express as px
import cv2
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from imantics import Polygons, Mask, BBox
from pycocotools import mask as maskUtils

from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet

lemmatizer = WordNetLemmatizer()

import warnings
warnings.filterwarnings("ignore")

from tqdm import tqdm
tqdm.pandas()

# COCO

In [2]:
f = open('coco/instances_val2017.json')
data = json.load(f)

f = open('coco/panoptic_val2017.json')
data_panoptic = json.load(f)

f = open('coco/stuff_val2017.json')
data_stuff = json.load(f)

In [3]:
new_data_panoptic = list()
for elem in data_panoptic["annotations"]:
    for seg in elem["segments_info"]:
        segment = seg
        segment["file_name"]=elem["file_name"]
        segment["image_id"]=elem["image_id"]
        segment["x1"]=segment["bbox"][0]
        segment["x2"]=segment["bbox"][1]
        segment["x3"]=segment["bbox"][2]
        segment["x4"]=segment["bbox"][3]
        new_data_panoptic.append(segment)

In [4]:
len(new_data_panoptic)

56728

In [5]:
df_panoptic = pd.DataFrame(new_data_panoptic)
df_panoptic

id  category_id  iscrowd                  bbox   area  \
0      3226956            1        0   [413, 158, 53, 138]   2840   
1      6979964            1        0    [384, 172, 16, 36]    439   
2      3103374           62        0    [413, 223, 30, 81]   1250   
3      2831194           62        0    [291, 218, 62, 98]   1848   
4      3496593           62        0    [412, 219, 10, 13]     90   
...        ...          ...      ...                   ...    ...   
56723  8703197           52        0  [467, 280, 173, 177]  22042   
56724  1938563           52        0  [137, 216, 297, 100]  13576   
56725  1995653           52        1     [0, 20, 640, 458]  24925   
56726  1325130          122        0      [0, 0, 640, 478]  83419   
56727  1252129          195        0      [0, 0, 578, 478]  52789   

              file_name  image_id   x1   x2   x3   x4  
0      000000000139.png       139  413  158   53  138  
1      000000000139.png       139  384  172   16   36  
2      000000000139.png       139  413  223   30   81  
3      000000000139.png       139  291  218   62   98  
4      000000000139.png       139  412  219   10   13  
...                 ...       ...  ...  ...  ...  ...  
56723  000000581781.png    581781  467  280  173  177  
56724  000000581781.png    581781  137  216  297  100  
56725  000000581781.png    581781    0   20  640  458  
56726  000000581781.png    581781    0    0  640  478  
56727  000000581781.png    581781    0    0  578  478  

[56728 rows x 11 columns]

In [6]:
new_data_stuff = list()
for elem in data_stuff["annotations"]:
    segment = elem
    segment["x1"]=segment["bbox"][0]
    segment["x2"]=segment["bbox"][1]
    segment["x3"]=segment["bbox"][2]
    segment["x4"]=segment["bbox"][3]
    new_data_stuff.append(segment)

In [7]:
df_stuff = pd.DataFrame(new_data_stuff).rename(columns={"category_id": "category_id_stuff", "id": "id_stuff", "bbox": "bbox_stuff"})
df_stuff = df_stuff[df_stuff["category_id_stuff"]!=183] # rimuovo classe other
df_stuff

segmentation      area  iscrowd  \
0      {'counts': 'j19[6h1ZNXNf1h1ZNYNe1g1[NYNf1f1YNZ...   25483.0        0   
1      {'counts': '`bh11Y=1O1O1O1O001O001O001O001O001...   20106.0        0   
2      {'counts': 'omh51Y=0ng31PXL01O10iW10ThN1PPY2',...      11.0        0   
3      {'counts': '_:k2_:00000O1000000000000O10000000...   49754.0        0   
4      {'counts': 'Qi\4:k<6O1bCAo;`0kC@M1W<n0O0N12O1O...     842.0        0   
...                                                  ...       ...      ...   
32794  {'counts': '0Y1l>l3O10OO2M300N101O1N101O1N101N...  176577.0        0   
32796  {'counts': '0[6c0\Lf4e3ZKZLf4b1mJfM>g0e4U1^KQN...   83419.0        0   
32797  {'counts': '[6c0\>0O1O100O2N1O101N2N1O2N1O010O...    1167.0        0   
32798  {'counts': 'e9T4j:O1YL[ET3Y;N1O10001O000000010...   52789.0        0   
32799  {'counts': 'cVX64j>2N0[AL_>5_AO_>702O0O2aB@^<`...    1473.0        0   

       image_id                  bbox_stuff  category_id_stuff  id_stuff  \
0           139   [0.0, 38.0, 549.0, 297.0]                 98  20000000   
1           139  [136.0, 0.0, 473.0, 116.0]                102  20000001   
2           139  [444.0, 226.0, 20.0, 11.0]                105  20000002   
3           139  [0.0, 269.0, 564.0, 157.0]                118  20000003   
4           139  [338.0, 166.0, 29.0, 50.0]                119  20000004   
...         ...                         ...                ...       ...   
32794    581615    [0.0, 0.0, 478.0, 640.0]                176  20032794   
32796    581781    [0.0, 0.0, 640.0, 478.0]                122  20032796   
32797    581781    [0.0, 203.0, 40.0, 40.0]                123  20032797   
32798    581781    [0.0, 0.0, 578.0, 478.0]                139  20032798   
32799    581781  [428.0, 407.0, 51.0, 71.0]                143  20032799   

          x1     x2     x3     x4  
0        0.0   38.0  549.0  297.0  
1      136.0    0.0  473.0  116.0  
2      444.0  226.0   20.0   11.0  
3        0.0  269.0  564.0  157.0  
4      338.0  166.0   29.0   50.0  
...      ...    ...    ...    ...  
32794    0.0    0.0  478.0  640.0  
32796    0.0    0.0  640.0  478.0  
32797    0.0  203.0   40.0   40.0  
32798    0.0    0.0  578.0  478.0  
32799  428.0  407.0   51.0   71.0  

[27849 rows x 11 columns]

In [8]:
df_things = pd.DataFrame(data["annotations"])
df_things["file_name"]=df_things['image_id'].apply(lambda x: '{0:0>16}'.format(str(x)+".jpg"))
df_things[['x1','x2', 'x3', 'x4']] = pd.DataFrame(df_things.bbox.tolist(), index= df_things.index)
df_things = df_things.sort_values(by=['image_id',"area"])

df_things

segmentation         area  \
27924  [[413.7, 220.47, 412.25, 231.06, 419.96, 231.5...     90.98725   
28442  [[339.52, 201.72, 336.79, 216.23, 346.35, 214....    120.23200   
27503  [[361.37, 229.69, 361.56, 226.09, 360.42, 220....    178.18510   
28441  [[242.95, 212.06, 241.24, 199.54, 254.32, 194....    189.56010   
24602  [[317.4, 219.24, 319.8, 230.83, 338.98, 230.03...    210.14820   
...                                                  ...          ...   
19107  [[140.71, 217.49, 147.16, 216.41, 201.94, 236....  13574.45810   
19164  [[40.91, 197.01, 37.68, 228.23, 3.23, 305.75, ...  14836.26035   
19176  [[587.56, 253.32, 556.41, 260.84, 540.3, 266.2...  16690.94945   
19151  [[532.91, 296.6, 570.59, 305.21, 589.96, 304.1...  22016.99120   
36740  {'counts': [468, 6, 444, 6, 13, 4, 1, 12, 1, 1...  28911.00000   

       iscrowd  image_id                              bbox  category_id  \
27924        0       139     [412.25, 219.02, 9.63, 12.52]           62   
28442        0       139      [336.79, 199.5, 9.73, 16.73]           86   
27503        0       139    [350.76, 208.84, 11.37, 22.55]           86   
28441        0       139    [241.24, 194.99, 14.22, 17.63]           86   
24602        0       139     [317.4, 219.24, 21.58, 11.59]           62   
...        ...       ...                               ...          ...   
19107        0    581781    [136.42, 216.41, 297.54, 99.9]           52   
19164        0    581781     [2.15, 184.09, 137.8, 159.34]           52   
19176        0    581781   [439.33, 94.35, 160.05, 171.86]           52   
19151        0    581781  [467.23, 280.45, 172.77, 177.63]           52   
36740        1    581781                 [0, 20, 639, 457]           52   

                 id         file_name      x1      x2      x3      x4  
27924       1941808  000000000139.jpg  412.25  219.02    9.63   12.52  
28442       2146548  000000000139.jpg  336.79  199.50    9.73   16.73  
27503       1669970  000000000139.jpg  350.76  208.84   11.37   22.55  
28441       2146194  000000000139.jpg  241.24  194.99   14.22   17.63  
24602        110334  000000000139.jpg  317.40  219.24   21.58   11.59  
...             ...               ...     ...     ...     ...     ...  
19107       1544126  000000581781.jpg  136.42  216.41  297.54   99.90  
19164       1547489  000000581781.jpg    2.15  184.09  137.80  159.34  
19176       1547752  000000581781.jpg  439.33   94.35  160.05  171.86  
19151       1546551  000000581781.jpg  467.23  280.45  172.77  177.63  
36740  905200581781  000000581781.jpg    0.00   20.00  639.00  457.00  

[36781 rows x 12 columns]

In [9]:
# final df 
df_coco = pd.merge(df_panoptic, df_stuff, on=['image_id', "x1", "x2", "x3", "x4"])[["id","category_id", "bbox","area_x", "file_name", "image_id", "segmentation", "id_stuff", "category_id_stuff","x1", "x2", "x3", "x4"]].rename(columns={"area_x": "area"})
df_coco["file_name"] = df_coco["file_name"].str.replace(".png",".jpg")
df_coco = pd.concat([df_coco, df_things]).sort_values(by=['image_id',"category_id"])
df_coco = df_coco.drop_duplicates(subset=["category_id", "image_id", "x1", "x2", "x3", "x4"])[["id", "category_id", "bbox", "area", "file_name", "image_id", "segmentation"]]

df_coco

id  category_id                              bbox  \
24776        233201            1    [384.43, 172.21, 15.12, 35.74]   
24772        230831            1    [412.8, 157.61, 53.05, 138.01]   
27924       1941808           62     [412.25, 219.02, 9.63, 12.52]   
24602        110334           62     [317.4, 219.24, 21.58, 11.59]   
24577        105328           62     [413.2, 223.01, 30.17, 81.36]   
...             ...          ...                               ...   
19176       1547752           52   [439.33, 94.35, 160.05, 171.86]   
19151       1546551           52  [467.23, 280.45, 172.77, 177.63]   
36740  905200581781           52                 [0, 20, 639, 457]   
19097       1325130          122                  [0, 0, 640, 478]   
19098       1252129          195                  [0, 0, 578, 478]   

              area         file_name  image_id  \
24776    435.14495  000000000139.jpg       139   
24772   2913.11040  000000000139.jpg       139   
27924     90.98725  000000000139.jpg       139   
24602    210.14820  000000000139.jpg       139   
24577   1289.37345  000000000139.jpg       139   
...            ...               ...       ...   
19176  16690.94945  000000581781.jpg    581781   
19151  22016.99120  000000581781.jpg    581781   
36740  28911.00000  000000581781.jpg    581781   
19097  83419.00000  000000581781.jpg    581781   
19098  52789.00000  000000581781.jpg    581781   

                                            segmentation  
24776  [[384.98, 206.58, 384.43, 199.98, 385.25, 193....  
24772  [[428.19, 219.47, 430.94, 209.57, 430.39, 210....  
27924  [[413.7, 220.47, 412.25, 231.06, 419.96, 231.5...  
24602  [[317.4, 219.24, 319.8, 230.83, 338.98, 230.03...  
24577  [[436.06, 304.37, 443.37, 300.71, 436.97, 261....  
...                                                  ...  
19176  [[587.56, 253.32, 556.41, 260.84, 540.3, 266.2...  
19151  [[532.91, 296.6, 570.59, 305.21, 589.96, 304.1...  
36740  {'counts': [468, 6, 444, 6, 13, 4, 1, 12, 1, 1...  
19097  {'counts': '0[6c0\Lf4e3ZKZLf4b1mJfM>g0e4U1^KQN...  
19098  {'counts': 'e9T4j:O1YL[ET3Y;N1O10001O000000010...  

[55849 rows x 7 columns]

In [10]:
len(set(df_coco["file_name"]))

4999

In [11]:
f = open('../coco/coco_categories_final.json')
 
df_coco_categories_final = pd.DataFrame(json.load(f))
df_coco_categories_final

supercategory   id           name   count  isthing  \
0          person    1         person  262351      NaN   
1            wall  199           wall   43524      0.0   
2         vehicle    3            car   43510      NaN   
3           plant  184           tree   41075      0.0   
4             sky  187            sky   40119      0.0   
..            ...  ...            ...     ...      ...   
128    structural  138            net    1362      0.0   
129       outdoor   14  parking meter    1285      NaN   
130        animal   23           bear    1282      NaN   
131     appliance   80        toaster     224      NaN   
132        indoor   89     hair dryer     198      NaN   

                                        original_class dataset  
0    {'original_class': 'person', 'original_id': 1,...    COCO  
1    {'original_class': 'wall-other-merged', 'origi...    COCO  
2    {'original_class': 'car', 'original_id': 3, 'd...    COCO  
3    {'original_class': 'tree-merged', 'original_id...    COCO  
4    {'original_class': 'sky-other-merged', 'origin...    COCO  
..                                                 ...     ...  
128  {'original_class': 'net', 'original_id': 138, ...    COCO  
129  {'original_class': 'parking meter', 'original_...    COCO  
130  {'original_class': 'bear', 'original_id': 23, ...    COCO  
131  {'original_class': 'toaster', 'original_id': 8...    COCO  
132  {'original_class': 'hair drier', 'original_id'...    COCO  

[133 rows x 7 columns]

In [12]:
df_coco = pd.merge(df_coco_categories_final.rename(columns={"id": "category_id"}), df_coco, on="category_id")
df_coco

supercategory  category_id        name   count  isthing  \
0            person            1      person  262351      NaN   
1            person            1      person  262351      NaN   
2            person            1      person  262351      NaN   
3            person            1      person  262351      NaN   
4            person            1      person  262351      NaN   
...             ...          ...         ...     ...      ...   
55844        indoor           89  hair dryer     198      NaN   
55845        indoor           89  hair dryer     198      NaN   
55846        indoor           89  hair dryer     198      NaN   
55847        indoor           89  hair dryer     198      NaN   
55848        indoor           89  hair dryer     198      NaN   

                                          original_class dataset       id  \
0      {'original_class': 'person', 'original_id': 1,...    COCO   233201   
1      {'original_class': 'person', 'original_id': 1,...    COCO   230831   
2      {'original_class': 'person', 'original_id': 1,...    COCO   442619   
3      {'original_class': 'person', 'original_id': 1,...    COCO   559508   
4      {'original_class': 'person', 'original_id': 1,...    COCO   560228   
...                                                  ...     ...      ...   
55844  {'original_class': 'hair drier', 'original_id'...    COCO  1676045   
55845  {'original_class': 'hair drier', 'original_id'...    COCO  1188212   
55846  {'original_class': 'hair drier', 'original_id'...    COCO  1675983   
55847  {'original_class': 'hair drier', 'original_id'...    COCO  2197655   
55848  {'original_class': 'hair drier', 'original_id'...    COCO  1188281   

                                   bbox         area         file_name  \
0        [384.43, 172.21, 15.12, 35.74]    435.14495  000000000139.jpg   
1        [412.8, 157.61, 53.05, 138.01]   2913.11040  000000000139.jpg   
2        [280.79, 44.73, 218.7, 346.68]  27789.11055  000000000785.jpg   
3      [145.26, 100.67, 291.95, 457.35]  25759.04240  000000000872.jpg   
4       [163.73, 126.42, 265.69, 480.4]  48091.75195  000000000872.jpg   
...                                 ...          ...               ...   
55844     [0.0, 101.67, 163.96, 181.22]  22528.82730  000000350002.jpg   
55845   [315.51, 187.84, 83.23, 226.16]   5066.24570  000000384136.jpg   
55846     [139.83, 48.57, 28.71, 38.29]    592.16200  000000500464.jpg   
55847     [155.44, 24.76, 34.97, 20.73]    302.02640  000000501368.jpg   
55848  [194.58, 209.63, 144.71, 229.02]  10347.80605  000000569976.jpg   

       image_id                                       segmentation  
0           139  [[384.98, 206.58, 384.43, 199.98, 385.25, 193....  
1           139  [[428.19, 219.47, 430.94, 209.57, 430.39, 210....  
2           785  [[353.37, 67.65, 358.15, 52.37, 362.92, 47.59,...  
3           872  [[310.65, 112.18, 339.42, 100.67, 362.43, 106....  
4           872  [[233.87, 606.82, 187.55, 602.85, 220.64, 580....  
...         ...                                                ...  
55844    350002  [[48.9, 274.26, 119.37, 282.89, 163.96, 242.62...  
55845    384136  [[386.71, 211.78, 375.66, 192.14, 352.95, 187....  
55846    500464  [[168.54, 86.47, 168.35, 86.86, 161.71, 85.69,...  
55847    501368  [[155.44, 37.72, 177.46, 24.76, 190.41, 26.06,...  
55848    569976  [[270.38, 336.54, 281.89, 320.72, 304.9, 340.8...  

[55849 rows x 13 columns]

In [13]:
len(set(df_coco["category_id"]))

133

In [14]:
df_coco["name"].value_counts()

person        11006
wall           2363
car            1932
chair          1791
sky            1611
              ...  
microwave        55
net              53
scissors         36
hair dryer       11
toaster           9
Name: name, Length: 127, dtype: int64

# LVIS

In [15]:
f = open('lvis/lvis_v1_val.json')
data_lvis = json.load(f)
df_lvis_annotations = pd.DataFrame(data_lvis["annotations"])
df_lvis_annotations

area      id                                       segmentation  \
0        73297.48       1  [[270.75, 598.57, 261.98, 598.57, 247.84, 598....   
1        70204.06       2  [[0.0, 626.21, 90.53, 629.18, 97.12, 611.7, 10...   
2          289.97       3  [[238.16, 386.84, 240.61, 387.95, 245.96, 385....   
3          238.39       4  [[392.16, 220.43, 395.79, 219.28, 400.38, 216....   
4       100734.18       5  [[23.83, 156.28, 0.0, 162.64, 0.0, 427.0, 166....   
...           ...     ...                                                ...   
244702       4.72  244703  [[209.04, 181.65, 208.15, 182.21, 207.92, 182....   
244703      15.41  244704  [[72.17, 253.4, 74.78, 254.09, 75.99, 252.7, 7...   
244704       2.75  244705  [[24.1, 257.12, 24.62, 257.47, 25.15, 257.64, ...   
244705       5.38  244706  [[215.52, 186.01, 214.45, 185.69, 213.38, 185....   
244706       8.27  244707  [[209.38, 180.15, 208.42, 180.06, 208.33, 179....   

        image_id                             bbox  category_id  
0         446522  [83.08, 219.88, 301.69, 420.12]          232  
1         446522       [0.0, 0.0, 132.41, 629.18]          421  
2          76261   [238.16, 340.87, 66.44, 47.08]           99  
3          76261   [368.84, 181.21, 51.42, 59.88]           99  
4         454750       [0.0, 14.7, 413.66, 412.3]         1202  
...          ...                              ...          ...  
244702    358717      [207.92, 181.65, 2.35, 2.8]          125  
244703    358717        [71.65, 249.5, 4.6, 4.59]          125  
244704    358717        [24.1, 256.0, 2.17, 1.64]          125  
244705    358717     [212.74, 185.69, 2.86, 2.46]          125  
244706    358717      [208.33, 177.29, 3.3, 2.95]          125  

[244707 rows x 6 columns]

In [16]:
len(set(df_lvis_annotations["image_id"]))

19626

In [17]:
f = open('../lvis/lvis_categories_final.json')
df_lvis_categories_final = pd.DataFrame(json.load(f))
df_lvis_categories_final

name  instance_count  \
0    banana           50552   
1      book           33353   
2    carrot           18049   
3     apple           17451   
4      pole           14276   
..      ...             ...   
586  animal             103   
587    sled             102   
588     van             102   
589   fruit             101   
590   clock             100   

                                                   def  \
0    elongated crescent-shaped yellow fruit with so...   
1    a written work or composition that has been pu...   
2    deep orange edible root of the cultivated carr...   
3    fruit with red or yellow or green skin and swe...   
4    a long (usually round) rod of wood or metal or...   
..                                                 ...   
586  a domesticated animal kept for companionship o...   
587  a vehicle or flat object for transportation ov...   
588  a recreational vehicle equipped for camping ou...   
589    any of numerous inedible fruits with hard rinds   
590                          a clock mounted on a wall   

                                        synonyms  image_count    id frequency  \
0                                       [banana]         1787    45         f   
1                                         [book]         1903   127         f   
2                                       [carrot]         1222   217         f   
3                                        [apple]         1207    12         f   
4                                   [pole, post]         1890   827         f   
..                                           ...          ...   ...       ...   
586                                        [pet]           79   794         c   
587                       [sled, sledge, sleigh]           56   970         c   
588  [camper_(vehicle), camping_bus, motor_home]           40   191         c   
589                                      [gourd]            6   509         r   
590                                 [wall_clock]           48  1154         c   

              synset                                     original_class  \
0        banana.n.02  {'original_class': 'banana', 'original_id': 45...   
1          book.n.01  {'original_class': 'book', 'original_id': 127,...   
2        carrot.n.01  {'original_class': 'carrot', 'original_id': 21...   
3         apple.n.01  {'original_class': 'apple', 'original_id': 12,...   
4          pole.n.01  {'original_class': 'pole', 'original_id': 827,...   
..               ...                                                ...   
586         pet.n.01  {'original_class': 'pet', 'original_id': 794, ...   
587        sled.n.01  {'original_class': 'sled', 'original_id': 970,...   
588      camper.n.02  {'original_class': 'camper_(vehicle)', 'origin...   
589       gourd.n.02  {'original_class': 'gourd', 'original_id': 509...   
590  wall_clock.n.01  {'original_class': 'wall_clock', 'original_id'...   

    dataset  
0      LVIS  
1      LVIS  
2      LVIS  
3      LVIS  
4      LVIS  
..      ...  
586    LVIS  
587    LVIS  
588    LVIS  
589    LVIS  
590    LVIS  

[591 rows x 10 columns]

In [18]:
len(set(df_lvis_categories_final["name"].values))

521

In [19]:
df_lvis = pd.merge(df_lvis_categories_final, df_lvis_annotations.rename(columns={"id": "id_annotations", "category_id": "id"}), on="id")
df_lvis

name  instance_count  \
0       banana           50552   
1       banana           50552   
2       banana           50552   
3       banana           50552   
4       banana           50552   
...        ...             ...   
240882   clock             100   
240883   clock             100   
240884   clock             100   
240885   clock             100   
240886   clock             100   

                                                      def      synonyms  \
0       elongated crescent-shaped yellow fruit with so...      [banana]   
1       elongated crescent-shaped yellow fruit with so...      [banana]   
2       elongated crescent-shaped yellow fruit with so...      [banana]   
3       elongated crescent-shaped yellow fruit with so...      [banana]   
4       elongated crescent-shaped yellow fruit with so...      [banana]   
...                                                   ...           ...   
240882                          a clock mounted on a wall  [wall_clock]   
240883                          a clock mounted on a wall  [wall_clock]   
240884                          a clock mounted on a wall  [wall_clock]   
240885                          a clock mounted on a wall  [wall_clock]   
240886                          a clock mounted on a wall  [wall_clock]   

        image_count    id frequency           synset  \
0              1787    45         f      banana.n.02   
1              1787    45         f      banana.n.02   
2              1787    45         f      banana.n.02   
3              1787    45         f      banana.n.02   
4              1787    45         f      banana.n.02   
...             ...   ...       ...              ...   
240882           48  1154         c  wall_clock.n.01   
240883           48  1154         c  wall_clock.n.01   
240884           48  1154         c  wall_clock.n.01   
240885           48  1154         c  wall_clock.n.01   
240886           48  1154         c  wall_clock.n.01   

                                           original_class dataset     area  \
0       {'original_class': 'banana', 'original_id': 45...    LVIS   216.73   
1       {'original_class': 'banana', 'original_id': 45...    LVIS   153.76   
2       {'original_class': 'banana', 'original_id': 45...    LVIS   366.66   
3       {'original_class': 'banana', 'original_id': 45...    LVIS   195.67   
4       {'original_class': 'banana', 'original_id': 45...    LVIS    24.90   
...                                                   ...     ...      ...   
240882  {'original_class': 'wall_clock', 'original_id'...    LVIS  4933.66   
240883  {'original_class': 'wall_clock', 'original_id'...    LVIS  1955.59   
240884  {'original_class': 'wall_clock', 'original_id'...    LVIS  3344.10   
240885  {'original_class': 'wall_clock', 'original_id'...    LVIS  3799.83   
240886  {'original_class': 'wall_clock', 'original_id'...    LVIS   174.49   

        id_annotations                                       segmentation  \
0                  558  [[347.16, 146.34, 348.62, 148.35, 348.62, 150....   
1                  559  [[454.44, 130.81, 458.22, 129.49, 461.43, 128....   
2                  560  [[357.26, 131.01, 355.56, 133.39, 356.58, 134....   
3                  561  [[437.8, 134.13, 441.01, 130.57, 444.93, 128.9...   
4                  562  [[308.01, 174.28, 307.26, 175.4, 306.7, 176.16...   
...                ...                                                ...   
240882          120495  [[201.58, 0.0, 201.16, 18.15, 199.46, 21.55, 2...   
240883          126976  [[496.41, 110.32, 500.47, 105.59, 503.04, 100....   
240884          130543  [[166.72, 44.26, 175.38, 43.07, 182.68, 41.71,...   
240885          130544  [[482.2, 122.5, 443.61, 121.3, 440.4, 116.75, ...   
240886          172071  [[306.18, 236.17, 308.64, 236.31, 311.24, 237....   

        image_id                            bbox  
0         480275  [344.41, 134.46, 21.78, 17.91]  
1         480275  [449.72, 128.35, 18.13, 14.72]  
2         48

In [20]:
len(set(df_lvis["name"]))

521

Sistemo COCO per eliminare le annotazioni duplicate

In [21]:
unique_lvis_classes = list(set(df_lvis["name"].values) - set(df_coco["name"].values))
print("numero di classi presenti solo in LVIS:",len(unique_lvis_classes))
unique_lvis_classes

numero di classi presenti solo in LVIS: 437


['wire',
 'earring',
 'crumb',
 'scooter',
 'bar',
 'bandana',
 'coffee table',
 'watermelon',
 'bookcase',
 'stirrup',
 'marker',
 'card',
 'steering wheel',
 'tablecloth',
 'box',
 'street lamp',
 'egg',
 'carton',
 'steak',
 'weather vane',
 'shirt',
 'can',
 'vent',
 'sweater',
 'wristband',
 'saddle blanket',
 'swimsuit',
 'cauliflower',
 'bag',
 'sunglasses',
 'dress',
 'date (fruit)',
 'short',
 'saddle',
 'stroller',
 'paper towel',
 'camera',
 'water tank',
 'button',
 'beer bottle',
 'plate',
 'cracker',
 'brake light',
 'buoy',
 'mattress',
 'blouse',
 'street light',
 'parasail',
 'pastry',
 'home plate',
 'croissant',
 'blender',
 'trunk',
 'jacket',
 'knee pad',
 'pajama',
 'cork (bottle plug)',
 'barrel',
 'ham',
 'shower curtain',
 'computer monitor',
 'mailbox',
 'belt buckle',
 'awning',
 'magazine',
 'mound',
 'panda',
 'ski jacket',
 'tortilla',
 'dish',
 'trash can',
 'jersey',
 'cooler',
 'pencil',
 'toothpick',
 'shoe',
 'whipped cream',
 'pennant',
 'clock tower

In [22]:
unique_coco_classes = list(set(df_coco["name"].values) - set(df_lvis["name"].values))
print("numero di classi presenti solo in COCO:",len(unique_coco_classes))
unique_coco_classes

numero di classi presenti solo in COCO: 43


['house',
 'furniture',
 'door',
 'road',
 'sky',
 'sand',
 'grass',
 'keyboard',
 'couch',
 'river',
 'tent',
 'rug',
 'shelf',
 'laptop',
 'wall',
 'plant',
 'rock',
 'pavement',
 'staircase',
 'net',
 'fence',
 'building',
 'mountain',
 'cell phone',
 'railroad',
 'water',
 'tree',
 'remote',
 'hot dog',
 'tie',
 'cardboard',
 'snow',
 'sea',
 'floor',
 'ceiling',
 'platform',
 'dirt',
 'bridge',
 'light',
 'roof',
 'window',
 'mouse',
 'gravel']

In [23]:
df_lvis_images = pd.DataFrame(data_lvis["images"]).rename(columns={"id":"image_id"})
df_lvis_images

date_captured                                   neg_category_ids  \
0      2013-11-14 17:02:52               [279, 899, 127, 180, 1136, 725, 663]   
1      2013-11-14 20:55:31  [1002, 434, 924, 928, 951, 378, 1012, 867, 108...   
2      2013-11-14 22:32:02  [194, 899, 248, 587, 201, 838, 1112, 1061, 337...   
3      2013-11-14 23:11:37                    [149, 713, 654, 1112, 753, 579]   
4      2013-11-14 23:16:55              [22, 369, 176, 653, 89, 11, 753, 474]   
...                    ...                                                ...   
19804  2013-11-24 13:20:54                [1014, 1148, 857, 80, 660, 142, 13]   
19805  2013-11-24 20:54:28    [782, 1126, 608, 857, 1053, 486, 380, 992, 270]   
19806  2013-11-24 21:59:45      [21, 857, 1008, 57, 800, 531, 15, 1202, 1120]   
19807  2013-11-24 22:39:24                         [335, 1085, 672, 190, 489]   
19808  2013-11-25 08:47:32  [452, 1166, 455, 738, 1151, 315, 1013, 1137, 3...   

       image_id  license  height  width  \
0        397133        4     427    640   
1         37777        1     230    352   
2        252219        4     428    640   
3         87038        1     480    640   
4        174482        6     388    640   
...         ...      ...     ...    ...   
19804    125910        4     640    480   
19805    342307        3     375    500   
19806    432062        1     433    640   
19807    347405        3     427    640   
19808    155379        6     480    640   

                                              flickr_url  \
0      http://farm7.staticflickr.com/6116/6255196340_...   
1      http://farm9.staticflickr.com/8429/7839199426_...   
2      http://farm4.staticflickr.com/3446/3232237447_...   
3      http://farm8.staticflickr.com/7355/8825114508_...   
4      http://farm8.staticflickr.com/7020/6478877255_...   
...                                                  ...   
19804  http://farm8.staticflickr.com/7392/9308795953_...   
19805  http://farm4.staticflickr.com/3076/3160532038_...   
19806  http://farm4.staticflickr.com/3114/2305083105_...   
19807  http://farm6.staticflickr.com/5511/9585778258_...   
19808  http://farm4.staticflickr.com/3760/9203741571_...   

                                                coco_url  \
0      http://images.cocodataset.org/val2017/00000039...   
1      http://images.cocodataset.org/val2017/00000003...   
2      http://images.cocodataset.org/val2017/00000025...   
3      http://images.cocodataset.org/val2017/00000008...   
4      http://images.cocodataset.org/val2017/00000017...   
...                                                  ...   
19804  http://images.cocodataset.org/train2017/000000...   
19805  http://images.cocodataset.org/train2017/000000...   
19806  http://images.cocodataset.org/train2017/000000...   
19807  http://images.cocodataset.org/train2017/000000...   
19808  http://images.cocodataset.org/train2017/000000...   

      not_exhaustive_category_ids  
0      [914, 801, 566, 139, 1021]  
1                       [181, 45]  
2                           [948]  
3                      [94, 1043]  
4                          [1112]  
...                           ...  
19804                      [1122]  
19805                          []  
19806                          []  
19807                          []  
19808                       [904]  

[19809 rows x 9 columns]

In [24]:
df_lvis = pd.merge(df_lvis, df_lvis_images, on="image_id")
df_lvis

name  instance_count  \
0       banana           50552   
1       banana           50552   
2       banana           50552   
3       banana           50552   
4       banana           50552   
...        ...             ...   
240882     van             102   
240883     van             102   
240884     van             102   
240885   clock             100   
240886   clock             100   

                                                      def  \
0       elongated crescent-shaped yellow fruit with so...   
1       elongated crescent-shaped yellow fruit with so...   
2       elongated crescent-shaped yellow fruit with so...   
3       elongated crescent-shaped yellow fruit with so...   
4       elongated crescent-shaped yellow fruit with so...   
...                                                   ...   
240882  a recreational vehicle equipped for camping ou...   
240883  a recreational vehicle equipped for camping ou...   
240884  a recreational vehicle equipped for camping ou...   
240885                          a clock mounted on a wall   
240886                          a clock mounted on a wall   

                                           synonyms  image_count    id  \
0                                          [banana]         1787    45   
1                                          [banana]         1787    45   
2                                          [banana]         1787    45   
3                                          [banana]         1787    45   
4                                          [banana]         1787    45   
...                                             ...          ...   ...   
240882  [camper_(vehicle), camping_bus, motor_home]           40   191   
240883  [camper_(vehicle), camping_bus, motor_home]           40   191   
240884  [camper_(vehicle), camping_bus, motor_home]           40   191   
240885                                 [wall_clock]           48  1154   
240886                                 [wall_clock]           48  1154   

       frequency           synset  \
0              f      banana.n.02   
1              f      banana.n.02   
2              f      banana.n.02   
3              f      banana.n.02   
4              f      banana.n.02   
...          ...              ...   
240882         c      camper.n.02   
240883         c      camper.n.02   
240884         c      camper.n.02   
240885         c  wall_clock.n.01   
240886         c  wall_clock.n.01   

                                           original_class dataset  ...  \
0       {'original_class': 'banana', 'original_id': 45...    LVIS  ...   
1       {'original_class': 'banana', 'original_id': 45...    LVIS  ...   
2       {'original_class': 'banana', 'original_id': 45...    LVIS  ...   
3       {'original_class': 'banana', 'original_id': 45...    LVIS  ...   
4       {'original_class': 'banana', 'original_id': 45...    LVIS  ...   
...                                                   ...     ...  ...   
240882  {'original_class': 'camper_(vehicle)', 'origin...    LVIS  ...   
240883  {'original_class': 'camper_(vehicle)', 'origin...    LVIS  ...   
240884  {'original_class': 'camper_(vehicle)', 'origin...    LVIS  ...   
240885  {'original_class': 'wall_clock', 'original_id'...    LVIS  ...   
240886  {'original_class': 'wall_clock', 'original_id'...    LVIS  ...   

        image_id                             bbox        date_captured  \
0         480275   [344.41, 134.46, 21.78, 17.91]  2013-11-19 18:49:01   
1         480275   [449.72, 128.35, 18.13, 14.72]  2013-11-19 18:49:01   
2         480275   [355.56, 130.84, 41.34, 16.32]  2013-11-19 18:49:01   
3         480275   [435.31, 127.72, 18.53, 20.48]  2013-11-19 18:49:01   
4         480275    [297.86, 171.27, 10.15, 4.89]  2013-11-19 18:49:01   
...          ...                              ...                  ...   
240882    521221   [110.08, 33.53, 156.22, 59.73]  2013-11-15 21:57:58   
240883    521221      [44.06, 45.0, 47.69, 22.51]  20

In [25]:
def extract_coco_filename(url):
    return url[-16:]

In [26]:
df_lvis["file_name"] = df_lvis["coco_url"].apply(extract_coco_filename)

In [27]:
len(set(df_lvis["file_name"]))

19548

In [28]:
coco_filename_not_in_lvis = list(set(df_coco["file_name"].values) - set(df_lvis["file_name"].values))
print(len(coco_filename_not_in_lvis))

278


In [29]:
lvis_filename_not_in_coco = list(set(df_lvis["file_name"].values) - set(df_coco["file_name"].values))
print(len(lvis_filename_not_in_coco))

14827


In [30]:
# Filtro COCO prendendo le classi non presenti in LVIS
df_coco = df_coco[(df_coco["name"].isin(unique_coco_classes))] #| (df_coco["file_name"].isin(coco_filename_not_in_lvis))]
len(set(df_coco["name"].values))

43

In [31]:
df_coco["name"].value_counts()

wall          2363
sky           1611
tree          1407
building       978
grass          955
floor          943
pavement       873
road           651
window         602
fence          554
light          448
dirt           444
ceiling        415
door           352
plant          343
sea            291
remote         283
house          277
rug            266
cell phone     262
couch          261
tie            254
mountain       247
laptop         231
cardboard      229
shelf          225
sand           212
rock           178
snow           174
roof           169
furniture      162
keyboard       153
hot dog        127
railroad       119
platform       113
gravel         113
mouse          106
staircase       92
river           88
bridge          87
water           79
tent            67
net             53
Name: name, dtype: int64

In [32]:
df_lvis["name"].value_counts()

banana      9156
book        7022
carrot      3762
apple       3116
pole        2913
            ... 
cube           7
eggplant       7
sled           7
windmill       5
candy          3
Name: name, Length: 521, dtype: int64

In [33]:
#df_lvis[df_lvis["name"]=="person"]

# ADE20K

In [34]:
f = open('./ade20k/ade_validation.json')
df_ade = pd.DataFrame(json.load(f))
df_ade

id_annotations        name  \
0                   0         sky   
1                   1        tree   
2                   2        wall   
3                   3    building   
4                   4      dormer   
...               ...         ...   
70862              15       cliff   
70863               1       stair   
70864               2    platform   
70865               3        wall   
70866               4  grandstand   

                                          original_class dataset  \
0      {'original_class': 'sky', 'original_id': 0, 'd...  ADE20K   
1      {'original_class': 'trees', 'original_id': 1, ...  ADE20K   
2      {'original_class': 'wall', 'original_id': 2, '...  ADE20K   
3      {'original_class': 'building', 'original_id': ...  ADE20K   
4      {'original_class': 'dormer', 'original_id': 4,...  ADE20K   
...                                                  ...     ...   
70862  {'original_class': 'cliff', 'original_id': 15,...  ADE20K   
70863  {'original_class': 'steps', 'original_id': 1, ...  ADE20K   
70864  {'original_class': 'platform', 'original_id': ...  ADE20K   
70865  {'original_class': 'wall', 'original_id': 3, '...  ADE20K   
70866  {'original_class': 'grandstand', 'original_id'...  ADE20K   

                       bbox             file_name              image_id  \
0         [0, 0, 1278, 675]  ADE_val_00000001.jpg  ADE_val_00000001.jpg   
1        [2, 577, 323, 735]  ADE_val_00000001.jpg  ADE_val_00000001.jpg   
2        [0, 718, 654, 816]  ADE_val_00000001.jpg  ADE_val_00000001.jpg   
3      [312, 87, 1275, 807]  ADE_val_00000001.jpg  ADE_val_00000001.jpg   
4      [802, 477, 915, 561]  ADE_val_00000001.jpg  ADE_val_00000001.jpg   
...                     ...                   ...                   ...   
70862    [163, 0, 255, 256]  ADE_val_00001999.jpg  ADE_val_00001999.jpg   
70863  [135, 193, 189, 239]  ADE_val_00002000.jpg  ADE_val_00002000.jpg   
70864   [50, 165, 351, 261]  ADE_val_00002000.jpg  ADE_val_00002000.jpg   
70865    [50, 38, 175, 181]  ADE_val_00002000.jpg  ADE_val_00002000.jpg   
70866   [172, 37, 351, 172]  ADE_val_00002000.jpg  ADE_val_00002000.jpg   

                                            segmentation  
0      [0, 0, 0, 588, 37, 588, 65, 588, 81, 604, 90, ...  
1      [3, 683, 2, 584, 10, 583, 30, 583, 45, 577, 52...  
2      [3, 816, 119, 808, 244, 809, 339, 801, 390, 80...  
3      [760, 783, 654, 797, 653, 724, 322, 713, 312, ...  
4      [861, 549, 915, 487, 854, 477, 802, 490, 808, ...  
...                                                  ...  
70862  [255, 0, 212, 0, 209, 2, 208, 8, 206, 9, 202, ...  
70863  [162, 239, 189, 226, 185, 200, 158, 193, 153, ...  
70864  [50, 261, 51, 182, 78, 170, 80, 192, 225, 233,...  
70865  [50, 38, 172, 38, 175, 124, 175, 130, 175, 136...  
70866  [349, 38, 172, 37, 175, 109, 181, 111, 181, 13...  

[70867 rows x 8 columns]

In [35]:
len(set(df_ade["file_name"]))

2000

# Union

In [36]:
df_coco = df_coco[["id", "name", "original_class", "image_id", "dataset", "bbox", "segmentation", "file_name"]].rename(columns={"id": "id_annotations"})
df_ade = df_ade[["id_annotations", "name", "original_class", "image_id", "dataset", "bbox", "segmentation", "file_name"]]
df_lvis = df_lvis[["id_annotations", "name", "original_class", "image_id", "dataset", "bbox", "segmentation", "file_name"]]
df_dataset_full = df_coco.append(df_lvis).append(df_ade)
df_dataset_full.groupby("dataset").size()

dataset
ADE20K     70867
COCO       17857
LVIS      240887
dtype: int64

In [37]:
f = open('../final_classes/final_vocabulary.json')
df_final_vocabulary = pd.DataFrame(json.load(f))[["class", "id"]]
df_dataset_full = pd.merge(df_dataset_full.rename(columns={"name": "class"}), df_final_vocabulary, on=["class"])
df_dataset_full["split"] = "validation"
df_dataset_full = df_dataset_full[["id_annotations", "class", "id", "dataset", "original_class", "image_id", "file_name", "bbox", "segmentation", "split"]]
df_dataset_full

id_annotations       class   id dataset  \
0              4224910        wall    5    COCO   
1              7897479        wall    5    COCO   
2              4074010        wall    5    COCO   
3              2637857        wall    5    COCO   
4              2697775        wall    5    COCO   
...                ...         ...  ...     ...   
322757              13  decoration  387  ADE20K   
322758              16  decoration  387  ADE20K   
322759              27  decoration  387  ADE20K   
322760               3        pond  497  ADE20K   
322761               6        pond  497  ADE20K   

                                           original_class  \
0       {'original_class': 'wall-other-merged', 'origi...   
1       {'original_class': 'wall-other-merged', 'origi...   
2       {'original_class': 'wall-other-merged', 'origi...   
3       {'original_class': 'wall-other-merged', 'origi...   
4       {'original_class': 'wall-other-merged', 'origi...   
...                                                   ...   
322757  {'original_class': 'decoration', 'original_id'...   
322758  {'original_class': 'decoration', 'original_id'...   
322759  {'original_class': 'decoration', 'original_id'...   
322760  {'original_class': 'pond', 'original_id': 3, '...   
322761  {'original_class': 'pond', 'original_id': 6, '...   

                    image_id             file_name                  bbox  \
0                        139      000000000139.jpg      [0, 0, 640, 358]   
1                        802      000000000802.jpg      [0, 0, 424, 640]   
2                        885      000000000885.jpg      [0, 0, 640, 255]   
3                       1000      000000001000.jpg   [342, 32, 254, 207]   
4                       1268      000000001268.jpg   [387, 268, 157, 59]   
...                      ...                   ...                   ...   
322757  ADE_val_00001477.jpg  ADE_val_00001477.jpg      [89, 1, 132, 48]   
322758  ADE_val_00001529.jpg  ADE_val_00001529.jpg  [118, 135, 131, 151]   
322759  ADE_val_00001859.jpg  ADE_val_00001859.jpg     [11, 7, 715, 312]   
322760  ADE_val_00001621.jpg  ADE_val_00001621.jpg    [1, 506, 941, 958]   
322761  ADE_val_00001666.jpg  ADE_val_00001666.jpg    [90, 93, 182, 131]   

                                             segmentation       split  
0       {'counts': '0j19a4c4_K]Ka4c4_K^K`4b4`K^Ka4a4^K...  validation  
1       {'counts': '0P\98hWGb0a\OXObb0Z2iN<D=l^OlLb`0T...  validation  
2       {'counts': '0n7]5ZHdJ\7]5`HhJ_7X5^HkJb7U5^HkJb...  validation  
3       {'counts': 'Q^P52\=OTD2aNNY=1UD5POJY<1eD7dNM0O...  validation  
4       {'counts': 'meQ51Z=102@MWC04:^<=M3N20000O1O2O0...  validation  
...                                                   ...         ...  
322757  [99, 41, 89, 25, 89, 12, 91, 4, 117, 1, 127, 1...  validation  
322758  [120, 135, 118, 140, 119, 148, 123, 151, 130, ...  validation  
322759       [715, 88, 530, 7, 14, 11, 11, 102, 715, 312]  validation  
322760  [15, 587, 151, 566, 279, 551, 381, 530, 458, 5...  validation  
322761  [99, 131, 111, 128, 136, 125, 156, 117, 175, 1...  validation  

[322762 rows x 10 columns]

In [38]:
len(set(df_dataset_full["class"]))

662

In [39]:
set(df_dataset_full["class"].values) - set(df_final_vocabulary["class"].values)

set()

In [40]:
 set(df_final_vocabulary["class"].values) - set(df_dataset_full["class"].values)

{'bouquet', 'bush', 'hedge', 'shrub'}

check per vedere se ci sono classi al di fuori di quelle di partenza

In [41]:
f = open('../original_classes/VG-SGG-dicts-vgoi6-clipped.json')
 
original_classes = json.load(f)
print("original number of categories: ",len(original_classes["label_to_idx"]))

original number of categories:  1594


In [42]:
df_original_classes = pd.DataFrame({"class": original_classes["object_count"].keys(), "count": original_classes["object_count"].values()})
df_original_classes.head()

class  count
0  window  53848
1    tree  50339
2     man  47202
3  person  43031
4   shirt  38238

In [43]:
set(df_dataset_full["class"].values) - set(df_original_classes["class"].values)

set()

Dubbi:
1. Come mi comporto con l'oversampling effettuto sul training? riduco anche sul validation le classi ridotte?
2. Mancano 4 classi nel validation, sistemo o lascio così? (ho rispettato la divisione originale dei datasets, bouquet proviene da Lvis mentre le altre da ADE20K)
3. Lvis e COCO hanno le immagini del validation e del train "mischiate" lascio così o seguo la divisione di Lvis (19548 immagini) che ha più immagini  quella di COCO (4618 immagini) che ne ha di meno ?
4. Come mi comporto per il test? uso quello usato da Mask R-CNN? non c'è un benchmark usato per valutare i modelli che estraggono le features